presto-hive/src/main/java/com/facebook/presto/hive/util/MergingPageIterator.java

```diff
@@ -0,0 +1,190 @@
+        this.pageBuilder = new PageBuilder(types);
+
+        iterators.stream()
+                .map(PagePositionIterator::new)
+                .filter(PagePositionIterator::hasNext)
+                .forEach(queue::add);
+    }
+
+    @Override
+    protected Page computeNext()
+    {
+        while (!pageBuilder.isFull()) {
+            PagePositionIterator iterator = queue.poll();
+            if (iterator == null) {
+                break;
+            }
+
+            iterator.appendNextTo(pageBuilder);
+
+            if (iterator.hasNext()) {
+                queue.add(iterator);
+            }
+        }
+
+        if (pageBuilder.isEmpty()) {
+            return endOfData();
+        }
+
+        Page page = pageBuilder.build();
+        pageBuilder.reset();
+        return page;
+    }
+
+    @SuppressWarnings("ComparableImplementedButEqualsNotOverridden")
+    private class PagePositionIterator
+            implements Comparable<PagePositionIterator>
+    {
+        private final Iterator<Page> pages;
+        private PagePosition current;
+
+        public PagePositionIterator(Iterator<Page> pages)
+        {
+            this.pages = requireNonNull(pages, "pages is null");
+        }
+
+        public boolean hasNext()
+        {
+            load();
+            return current != null;
+        }
+
+        public void appendNextTo(PageBuilder pageBuilder)
+        {
+            ensureLoaded();
+            current.appendNextTo(pageBuilder);
+            if (!current.hasNext()) {
+                current = null;
+            }
+        }
+
+        @SuppressWarnings("CompareToUsesNonFinalVariable")
+        @Override
+        public int compareTo(PagePositionIterator other)
+        {
+            ensureLoaded();
+            other.ensureLoaded();
+            return current.compareTo(other.current);
+        }
+
+        private void ensureLoaded()
+        {
+            load();
+            checkState(current != null, "no more positions");
+        }
+
+        private void load()
+        {
+            while ((current == null) && pages.hasNext()) {
+                Page page = pages.next();
+                if (page.getPositionCount() != 0) {
+                    current = new PagePosition(page);
+                }
+            }
+        }
+    }
+
+    @SuppressWarnings("ComparableImplementedButEqualsNotOverridden")
+    private class PagePosition
+            implements Comparable<PagePosition>
+    {
+        private final Page page;
+        private int position;
+
+        public PagePosition(Page page)
+        {
+            this.page = requireNonNull(page, "page is null");
+        }
+
+        public boolean hasNext()
+        {
+            return position < page.getPositionCount();
+        }
+
+        public void appendNextTo(PageBuilder pageBuilder)
+        {
+            checkState(hasNext(), "no more positions");
+            appendPositionTo(page, position, pageBuilder);
+            position++;
```

Comment:
this seems like pre-optimization. If you would like to reuse `MergeHashSort` (via `WorkProcessor`) logic then it would be possible to reuse mutable `PageWithPosition` because next page with position is returned only when work processor is explicitly asked to.